packages/react-kit/src/components/Popover/Popover.tsx

```diff
@@ -430,27 +431,55 @@ type TVerticalPosition = {
 	align: PopoverAlign;
 };
 
-function movePopoverVertically(
-	placement: PopoverPlacement,
-	align: PopoverAlign,
-	anchorTop: number,
-	anchorBottom: number,
-	popoverHeight: number,
-	parentHeight: number,
+type TVerticalMoveProps = {
+	placement: PopoverPlacement;
+	align: PopoverAlign;
+	anchorTop: number;
+	anchorBottom: number;
+	popoverHeight: number;
+	parentHeight: number;
+	previousPlacement?: PopoverPlacement;
+	checkBounds?: boolean;
+};
+
+const movePopoverVertically: (props: TVerticalMoveProps) => TVerticalPosition | undefined = ({
+	placement,
+	align,
+	anchorTop,
+	anchorBottom,
+	popoverHeight,
+	parentHeight,
+	previousPlacement,
 	checkBounds = false,
-): TVerticalPosition | undefined {
+}) => {
 	switch (placement) {
 		case PopoverPlacement.Top: {
 			const top = anchorTop - popoverHeight;
+
 			if (checkBounds && top < 0) {
-				return movePopoverVertically(
-					PopoverPlacement.Bottom,
-					align,
-					anchorTop,
-					anchorBottom,
-					popoverHeight,
-					parentHeight,
-				);
+				switch (previousPlacement) {
+					case PopoverPlacement.Bottom:
+						return movePopoverVertically({
+							placement: PopoverPlacement.Bottom,
+							align,
+							anchorTop,
+							anchorBottom,
+							popoverHeight,
+							parentHeight,
+							previousPlacement: placement,
+						});
+					default:
+						return movePopoverVertically({
+							placement: PopoverPlacement.Bottom,
+							align,
+							anchorTop,
+							anchorBottom,
+							popoverHeight,
+							parentHeight,
+							previousPlacement: placement,
+							checkBounds: true,
```

Comment:
If I got it right, all the difference between cases it that if previous placement is not the same as current, we should checkBounds. But you have both `placement` and `previousPlacement` in arguments, maybe you could you test if you need check bounds inside of `movePopoverVertically`? Or, if we need to control `checkBounds` from outside, you could at least refactor this switch/case using `checkBounds: previousPlacement === PopoverPlacement.Bottom` (or whatever placement you need)

Reply:
What about tests guys?

Reply:
i moved methods of auto-positioning in Popover.model and wrote tests.